### PR TITLE
Linking file branch

### PR DESCRIPTION
**Pull request by williyamshoe**
Generates a linking.js, then in each html nightly file, references this file to create the previous, next, last, and first links. 
linking.js is generated within core.py, and can accommodate if a subset is given. This branch builds off of #54.

### CODE REVIEW

File: py/surveyqa/core.py [hunk 1]
```diff
@@ -50,6 +52,57 @@ def check_offline_files(dir):
...
+            inner_dict["prev"] = "night-"+f[i-1]+".html"
+            inner_dict["next"] = "night-"+f[i]+".html"
+        else:
+            inner_dict["prev"] = "night-"+f[i-1]+".html"
+            inner_dict["next"] = "night-"+f[i+1]+".html"
+        file_js["n"+f[i]] = inner_dict
+
+    outfile = os.path.join(outdir, 'linking.js')
+    with open(outfile, 'w') as fp:
+        fp.write("get_dict({})".format(str(file_js)))
```
Comment (sbailey) [inline py/surveyqa/core.py]: This seems a little magical to me.  Is it relying upon the fact that python dictionaries represented as strings are extremely similar to json?  I think they differ by single vs. double quote usage and browsers might be forgiving of that, but better to use `json.dumps(file_js)`.  Or maybe I'm completely missing how this works.

File: py/surveyqa/core.py [hunk 2]
```diff
@@ -99,6 +152,9 @@ def change_range(i):
     elif show_summary!="no":
         raise ValueError('show_summary should be "all", "subset", or "no". The value of show_summary was: {}'.format(show_summary))
 
+    nights_sub = sorted(set(exposures_sub['NIGHT']))
+    generate_js(outdir, nights_sub, nights != None)
```
Comment (sbailey) [inline py/surveyqa/core.py]: Please use a more descriptive function name than `generate_js`.  Generate js for what?

File: py/surveyqa/nightly.py [hunk 3]
```diff
@@ -126,6 +126,27 @@ def plot_timeseries(source, name, color, tools=None, x_range=None, title=None, t
...
+    '''Placeholder for a sky brightness timeseries plot'''
+    p = bk.figure(plot_width=width, plot_height=height,
+                  x_axis_label='Time',
+                  y_axis_label='Sky Brightness',
+                  min_border_left=min_border_left,
+                  min_border_right=min_border_right)
+    p.xaxis.axis_label_text_color='#ffffff'
+    p.toolbar_location = None
+    return p
+
```
Comment (sbailey) [inline py/surveyqa/nightly.py]: I think these two were removed by an earlier PR and are accidentally being added back in.  Double check that they aren't used anymore and remove from this PR if possible.

File: py/surveyqa/nightly.py [hunk 4]
```diff
@@ -496,6 +517,25 @@ def makeplots(night, exposures, tiles, outdir):
...
+#           <li><a id="last">Last</a></li>
+#           <li><a id="next">Next</a></li>
+#           <li><a id="prev">Previous</a></li>
+#           <li><a id="first">First</a></li>
+#           <li><a href={}>Summary Page</a></li>
+#         </ul>
+#     """.format(night, summary_str)
+
+# TODO: this is the link code from the earlier PR
+
```
Comment (sbailey) [inline py/surveyqa/nightly.py:517]: I resolved merge conflicts, but there is a "how should we do it" conceptual conflict with the previously merged PR #63 .  I left both in here and let's chat as a group on Monday for how to handle the navigation links for the first and last nights in light of this PR method.

File: py/surveyqa/nightly.py [hunk 5]
```diff
@@ -544,6 +584,22 @@ def makeplots(night, exposures, tiles, outdir):
             </ul>
         """.format(**navigation_links)
 
+# TODO: this is the code that updates the hrefs
+        
+    #- Update the navigation hrefs using the cached valus in linking.js
+    template += """
+    <script>
+        function get_dict(dict) {
```
Comment (sbailey) [inline py/surveyqa/nightly.py]: Please use a more descriptive name.   get dict for what?  Input dict is what?  It looks like this function is setting navigation elements, not getting a dictionary.  Please add comments in script about what is does, what the input should be, etc., similar to the type of information we include with python docstrings.

**Review by sbailey — CHANGES_REQUESTED** [review]
I added some comments inline.  Some of them about overly generic variable and function names could be addressed, and then let's chat as a group about how to handle the first/last night prev/next links with this PR compared to PR #63.

**Comment by williyamshoe** [comment]
I think I fixed all the issues above, and I have added the function where a link are un-clickable if there is no previous/next link.

**Review by sbailey — APPROVED** [review]

**Comment by sbailey** [comment]
Thanks.  Works well.  I renamed `generate_js_containing_json_file` to the slightly more informative (to me) name `write_night_linkage`.  i.e. trying to give a hint about the meaning of the contents (night linkage) rather than the format details (js containing json, which could be anything).

**Pull request merged**